cpp/ql/src/semmle/code/cpp/controlflow/StackVariableReachability.qll

```diff
@@ -3,7 +3,194 @@
+  ) {
+    this = e and
+    (
+      testIsTrue = true and condIsTrue = true
+      or
+      testIsTrue = false and condIsTrue = false
+    )
+  }
+}
+
+private class NotGuardCondition extends LogicalGuardCondition, NotExpr {
+  override predicate impliesCondition(
+    LogicalGuardCondition e, boolean testIsTrue, boolean condIsTrue
+  ) {
+    e = this.getOperand() and
+    (
+      testIsTrue = true and
+      condIsTrue = false
+      or
+      testIsTrue = false and
+      condIsTrue = true
+    )
+  }
+}
+
+/** A condition that either the trivial `True` condition, or a complex `LogicalCondition`. */
+abstract private class Condition extends TCondition {
+  abstract string toString();
+
+  /**
+   * Holds if this condition having the value `this.getTruthValue()` implies that `cond` has truth
+   * value `cond.getTruthValue()`.
+   */
+  pragma[nomagic]
+  abstract predicate impliesCondition(Condition cond);
+
+  /** Gets the value of this `Condition`. */
+  abstract boolean getTruthValue();
+
+  /** Gets the negated the expression represented by this `Condition`, if any. */
+  Condition negate() { none() }
+
+  /**
+   * Holds if this condition having the value `this.getTruthValue()` implies that `cond` cannot have
+   * the truth value `cond.getTruthValue()`.
+   */
+  final predicate refutesCondition(Condition cond) { this.impliesCondition(cond.negate()) }
+
+  /** Gets the `Location` of the expression that generated this `Condition`, if any. */
+  Location getLocation() { none() }
+}
+
+private class True extends Condition, TTrue {
+  True() { this = TTrue() }
+
+  override string toString() { result = "True" }
+
+  override boolean getTruthValue() { result = true }
+
+  override predicate impliesCondition(Condition cond) { cond instanceof True }
+}
+
+private class LogicalCondition extends Condition, TLogicalCondition {
+  boolean testIsTrue;
+  LogicalGuardCondition guard;
+
+  LogicalCondition() { this = TLogicalCondition(guard, testIsTrue) }
+
+  override string toString() { result = guard.toString() + " == " + testIsTrue.toString() }
+
+  override boolean getTruthValue() { result = testIsTrue }
+
+  LogicalGuardCondition getCondition() { result = guard }
+
+  override predicate impliesCondition(Condition cond) {
+    cond instanceof True
+    or
+    exists(LogicalGuardCondition other |
+      other = cond.(LogicalCondition).getCondition() and
+      this.getCondition()
+          .impliesCondition(globalValueNumber(other).getAnExpr(),
+            pragma[only_bind_into](pragma[only_bind_out](testIsTrue)),
+            pragma[only_bind_into](pragma[only_bind_out](cond.getTruthValue())))
+    )
+  }
+
+  override LogicalCondition negate() {
+    result.getCondition() = guard and
+    result.getTruthValue() = testIsTrue.booleanNot()
+  }
+
+  override Location getLocation() { result = guard.getLocation() }
+}
+
+/**
+ * Gets a `Condition` that controls `b`.
+ *
+ * Note: The trivial `True` condition always controls `b`.
```

Comment:
To check my understanding, this is a Condition that must hold to enter `b`?

Reply:
Correct.